front/lib/api/assistant/generation.ts

```diff
@@ -235,12 +238,15 @@ export async function renderConversationForModel({
     },
     "[ASSISTANT_TRACE] Genration message token counts for model conversation rendering"
   );
+  while (selected.length > 0 && selected[0].message.role === "agent") {
+    selected.shift();
+  }
 
   return new Ok({
     modelConversation: {
-      messages: selected,
+      messages: selected.map((s) => s.message),
     },
-    tokensUsed,
+    tokensUsed: selected.map((s) => s.tokenUsed).reduce((a, b) => a + b, 0),
```

Comment:
I thiiiink this calculation is incorect since `tokensUsed` is initialized with `let tokensUsed = promptCountRes.value + tokensMargin;` but also edited on l.203 without a selected.unshift.

I would rather update tokenUsed to be tokensUsed in `selected` and keep the latest snapshot and pull the tokensUsed of the last element of `selected` here?

Reply:
Indeed. I will follow @fontanierh 's advice on only decreasing it when I pop, which is more robust. Thanks for catching it.

Reply:
s/pop/shift/